knowledge-base/adr/0035-destination-ts-type.md

```diff
@@ -0,0 +1,108 @@
+# The ts type/interface of `Destination` 
+
+## Status
+
+Proposal: Use alternative A as an easy solution that works for the major use cases.
+
+## Context
+The `Destination` ts type is used in the following scenario:
+1. [connectivity] The return type of the `getDestination`.
+2. [http-client] A parameter of the `executeHttpRequest`.
+
+The definition of the `Destination` looks like:
+```ts
+interface Destination {
+  // This is a mandatory key
+  url: string;
+  name?: string | null;
+  type?: 'HTTP' | 'LDAP' | 'MAIL' | 'RFC';
+  ...
+}
+```
+
+However, when supporting destinations with type `MAIL`, the mandatory `url` key becomes a problem, as mail destination saves the host url in the properties of the destination.
+
+## Alternatives
+
+### A: Make `url` optional
```

Comment:
[pp] I would go for option A, since it seems the most comfortable and return the `Destination` type in the getDestination() method and allow also the `Destination` in the `executeHttpRequest`. I would create the types `HttpDestination` and `MailDestination` anyhow and provided some typeguards `isHttpDestination` and `isMailDestination` which are used in our code then to narrow the type within the `executeHttpRequest` so the internal interface so we good there and throw a exception if the wrong type is provided.

Since the response of `getDestination` depends on a service at runtime we do not get much from a static type check and need the runtime exception anywat. Therefore I would keep it simple in the user facing interface.

Reply:
But option `B` is also fine for me.

Reply:
`B` is a breaking change, unfortunately.